ui/raidboss/browser_tts_engine.js

```diff
@@ -0,0 +1,87 @@
+'use strict';
+
+const TTSEngineType = {
+  SpeechSynthesis: 0,
+  GoogleTTS: 1,
+};
+
+class TTSItem {
+  play() {}
+}
+
+class SpeechTTSItem extends TTSItem {
+  constructor(text) {
+    super();
+    this.text = text;
+    this.item = new SpeechSynthesisUtterance(text);
+  }
+
+  play() {
+    window.speechSynthesis.speak(this.item);
+  }
+}
+
+class GoogleTTSItem extends TTSItem {
+  constructor(text) {
+    super();
+    this.text = text;
+    let iframe = document.createElement('iframe');
+    // remove sandbox so we can modify contents/call play on audio element later
+    iframe.removeAttribute('sandbox');
+    iframe.style.display = 'none';
+    document.body.appendChild(iframe);
+    let encText = encodeURIComponent(text);
+    iframe.contentDocument.body.innerHTML = '<audio src="https://translate.google.com/translate_tts?ie=UTF-8&client=tw-ob&tl=en&q=' + encText + '" id="TTS">';
+    this.item = iframe.contentDocument.body.firstElementChild;
+  }
+
+  play() {
+    this.item.play();
+  }
+}
+
+class BrowserTTSEngine {
+  constructor() {
+    this.engineType = null;
+    this.ttsItems = {};
+  }
+
+  play(text) {
+    try {
+      if (this.ttsItems[text] !== undefined) {
+        this.ttsItems[text].play();
+      } else {
+        switch (this.engineType) {
+        case null:
+          // figure out what TTS engine type we need
+          if (window.speechSynthesis !== undefined &&
+            window.speechSynthesis.getVoices().length > 0)
```

Comment:
Oddly, I found this extremely inconsistent.  If I open up raidboss in Chrome 82.0.4073.0, click the button, and then type `window.speechSynthesis.getVoices()` in the console, it comes up empty and then the second time it comes up with a list of voices.  So, in practice, this never works out.  Maybe there's a bug in Chrome, but something is awry here for sure.  I'm also fine to merge this as-is and fix this later, as both paths seem to work just fine once the voices get initialized.

Reply:
So it seems like the voices list doesn't populate until after the first interaction with `speechSynthesis`, and even then it loads async. I'll think about how to better handle this, maybe default to the google service and fall back to browser tts based on if the event fires successfully and has voices afterwards...

Reply:
I've refactored this to a manner that _should_ work properly now with the latest changes, but I don't have a good way to test it on my end. Can anyone with a working `window.speechSynthesis` that has voices test the new logic? Any Chrome on Windows should work, just pop open the dev tools and breakpoint `browser_tts_engine.js` line 50 and see if it's ever hit on a page refresh.

Reply:
Works for me, thanks!